src/Bicep.Core/Semantics/ParamsSemanticModel.cs

```diff
@@ -27,11 +30,96 @@ public ParamsSemanticModel(BicepParamFile bicepParamFile, Compilation? bicepComp
+        /// </summary>
+        /// <returns></returns>
+        private IEnumerable<IDiagnostic> GetAdditionalSemanticDiagnostics()
+        {
+            if (this.BicepCompilation is null)
+            {
+                return Enumerable.Empty<IDiagnostic>();
+            }
+            var requiredParameters = this.BicepCompilation.GetEntrypointSemanticModel().Parameters;
+            var parameterAssignmentSyntax = BicepParamFile.ProgramSyntax.Children.OfType<ParameterAssignmentSyntax>();
+
+            var sortedRequiredParameters =  requiredParameters.OrderBy(x => x.Name);
+            var sortedParameterAssignments = parameterAssignmentSyntax.OrderBy(x => this.ParamBinder.GetSymbolInfo(x)?.Name);
+
+            // get errors relating to missing declarations or assignments
+            AddMissingErrors(sortedRequiredParameters.ToArray(), sortedParameterAssignments.ToArray());
+            // get errors relating to type mismatch of params between Bicep and params files
+            AddTypeMismatchErrors(sortedParameterAssignments);
+
+            return semanticDiagnostics;
+        }
+
+        private void AddMissingErrors(Metadata.ParameterMetadata[] sortedRequiredParameters, ParameterAssignmentSyntax[] sortedParameterAssignments)
+        {
+            var i = 0;
+            var j = 0;
+
+            while ((i < sortedRequiredParameters.Length) && (j < sortedParameterAssignments.Length))
+            {
+                var bicepParam = sortedRequiredParameters[i];
+                var paramParam = this.ParamBinder.GetSymbolInfo(sortedParameterAssignments[j]);
+                
+                if (LanguageConstants.IdentifierComparer.Equals(bicepParam.Name, paramParam?.Name))
+                {
+                    i++;
+                    j++;
+                }
+                else if (LanguageConstants.IdentifierComparer.Compare(bicepParam.Name, paramParam?.Name) < 0)
+                {
+                    if (bicepParam.IsRequired)
+                    {
+                        semanticDiagnostics.Add(DiagnosticBuilder.ForPosition(new TextSpan(0,0)).MissingParameterAssignment(bicepParam.Name));
+                    }
+                    i++;
+                }
+                else
+                {
+                    semanticDiagnostics.Add(DiagnosticBuilder.ForPosition(sortedParameterAssignments[j].Span).MissingParameterDeclaration(paramParam?.Name));
+                    j++;
+                }
+            }
+
+            // if i or j is at the end, emit errors for all remaining unpaired symbols 
+            while (i < sortedRequiredParameters.Count())
+            {
+                if (sortedRequiredParameters[i].IsRequired)
+                {
+                    semanticDiagnostics.Add(DiagnosticBuilder.ForPosition(new TextSpan(0,0)).MissingParameterAssignment(sortedRequiredParameters[i].Name));
```

Comment:
The ToListDiagnosticWriter class: https://github.com/Azure/bicep/blob/8e74bc9e42bc65c57cfd4518c321e3ca125e015a/src/Bicep.Core/Emit/EmitLimitationCalculator.cs#L25

Should allow you to simplify to:
```csharp
diagnosticWriter.Write(new TextSpan(0, 0), x => x.MissingParameterAssignment(sortedRequiredParameters[i].Name));
```